src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py

```diff
@@ -300,7 +300,73 @@ def minimize_corpus(self, target_path, arguments, input_dirs, output_dir,
     Returns:
       A FuzzResult object.
     """
-    raise NotImplementedError
+    runner = _get_runner(target_path)
+
+    # Step 1: Generate corpus file for Centipede.
+    full_corpus_workdir = self._create_temp_dir('full_corpus_workdir')
+    input_dirs_param = ','.join(str(dir) for dir in input_dirs)
+    args = [
+        f'--workdir={full_corpus_workdir}',
+        f'--binary={target_path}',
+        f'--corpus_dir={input_dirs_param}',
+        '--num_runs=0',
+    ]
+    result = runner.run_and_wait(additional_args=args, timeout=max_time)
+
+    if result.timed_out:
+      logs.log_error(
+          ('Corpus minimization timed out: Failed to generate Centipede corpus '
+           'file'),
+          fuzzer_output=result.output)
+      raise TimeoutError('Minimization timed out.')
+
+    # Step 2: Distill.
+    args = [
+        f'--workdir={full_corpus_workdir}',
+        f'--binary={target_path}',
+        '--distill',
+    ]
+    result = runner.run_and_wait(additional_args=args, timeout=max_time)
+
+    if result.timed_out:
+      logs.log_error(
+          'Corpus minimization timed out: Failed to distill',
+          fuzzer_output=result.output)
+      raise TimeoutError('Minimization corpus timed out.')
+
+    # Step 3: Generate corpus files for output_dir.
+    os.makedirs(output_dir, exist_ok=True)
+    minimized_corpus_workdir = self._create_temp_dir('minimized_corpus_workdir')
+    distilled_file = os.path.join(
+        full_corpus_workdir,
+        f'distilled-{os.path.basename(target_path)}.000000')
+    corpus_file = os.path.join(minimized_corpus_workdir, 'corpus.000000')
+    shutil.copyfile(distilled_file, corpus_file)
+
+    args = [
+        f'--workdir={minimized_corpus_workdir}',
+        f'--corpus_to_files={output_dir}',
+    ]
+    result = runner.run_and_wait(additional_args=args, timeout=max_time)
```

Comment:
One more comment: do we need to divide up `max_time` somehow bettween the multiple invocations? i.e. subtract the time passed as we invoke centipede again and again.

Reply:
This is a good point. I can divide the max_time between step 2 (distill) and step 3 (regenerate corpus).
However, I should extend the max_time in the first step, because that step needs to re-run all existing input testcases, which is very time-consuming. We cannot skip it because it reproduces the feature file required for distillation.

Reply:
Ah never mind, I just noticed the time allowed for corpus minimization is [22 hours](https://github.com/google/clusterfuzz/blob/6495567961d80ad8955c0e946c90a420bff6c5fc/src/clusterfuzz/_internal/bot/tasks/utasks/corpus_pruning_task.py#L60), which should be more than enough.
I will simply subtract the time consumed as you suggested.